Update NAT config page to address https://github.com/ipfs/ipfs-docs/issues/1424

ElPaisano opened the pull request:
# [Latest draft link](https://bafybeieoqz7uzhcghfu56wr4p7z53hvyutn3zvkxhoflyz5saqzgacfr24.on.fleek.co//how-to/nat-configuration/)

# Describe your changes
Addresses https://github.com/ipfs/ipfs-docs/issues/1424

# Files changed
docs/how-to/nat-configuration.md

# What issue(s) does this address?
 https://github.com/ipfs/ipfs-docs/issues/1424

# Does this update depend on any other PRs?
No

## Checklist before requesting a review
- [x] Passing the beta version of the **Check Markdown links for modified files** check. Action results can be viewed [here](https://github.com/ipfs/ipfs-docs/actions/workflows/action.yml).

## Checklist before merging
- [x] Passing all required checks (The beta **Check Markdown links for modified files** check is not required)

Jorropo commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,136 @@
 
-## Port forwarding
+<!--
+### Enable DCUtR Holepunching 
+
+Not sure what to put here
+
+Is there a recommended repo or libp2p docs link that we can link to here that explains how to set this up for Kubo? 
+
+https://github.com/libp2p/specs/blob/master/relay/DCUtR.md doesn't seem right
+
+https://docs.libp2p.io/concepts/nat/hole-punching/#phase-ii-hole-punching? seems to be mostly conceptual
+
+https://blog.ipfs.tech/2022-01-20-libp2p-hole-punching/ seems to just explain the theory
+
+Maybe https://docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html and tell the reader that they can follow this tutorial to set up DCUtR Holepunching using Rust? Not specific to Kubo but seems like its the best we've got right now?
+!-->
```
Dcutr is enabled by default, the user does not have to do anything.
I think we just need to explain what it is the draw backs and / or link to the blog post.

Some people have hole punchable nats so they see some success rate, however they would have better connectivity with a port forward.

ElPaisano replied:
Alright, based the info off of kubo/docs and the blog post 
https://bafybeidbxtj5a5t6ashyyaq2j23gy6q2muga2yaftbkx2xxlgjkebmax6u.on.fleek.co/how-to/nat-configuration/#disable-dcutr-holepunching

Jorropo commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,136 @@
+#### Open a port 
+
+:::tip
+Assuming you're not trying to expose your daemon's API to the public internet, opening port `4001`/`tcp` should be sufficient.
```
As a user, this sentence lead me to belive exposing the api port could be something i could want to do, which no, never, dont do that.

Jorropo commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,136 @@
 1. Enter the IP address of your IPFS node.
-1. Set traffic to go through port `4001`.
+1. Set traffic to go through an outside port. `4001` is recommended if you are unsure.
 1. Reboot your IPFS node for the changes to take effect. Make sure to reboot the entire machine, not just the IPFS daemon.
```
This is required later, after editing the config.

ElPaisano commented:
@Jorropo I think I addressed your feedback, please take a look at the latest https://bafybeidbxtj5a5t6ashyyaq2j23gy6q2muga2yaftbkx2xxlgjkebmax6u.on.fleek.co/how-to/nat-configuration/ and let me know

Jorropo commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,170 @@
+
+### Enable IPv6
 
 If your router and internet service provider (ISP) support IPv6, enabling it will mitigate some connection issues. Most modern routers have a single option to enable IPv6. We are unable to give detailed information about each router's settings and preferences here. Search your router manufacturer's website for _IPv6_ for information on how to enable it.
```
Add a disclaimer here IPv6 does not always solve everything, my isp have per machine ipv6 but it still run a firewall that blocks inbound connections.
I belive that is fairly commun.

Jorropo commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,170 @@
-## Port forwarding
 
-Each router has different options and solutions for port forwarding. Most router manufacturers have guides for setting up port forwarding on their devices. In general, the steps are something like this:
+### Disable DCUtR Holepunching 
```
You don't need to disable it, Kubo only turn it off if you are not reachable.
Doing a port forward will "automatically" disable this.

You don't need to disable it.
What I hope from this point is to explain that:
1. DCUtR have a 3~6s extra latency when connecting to a peer.
2. DCUtR does not always work.

And thus you probably want to seek a better alternative if you care about hosting content behind a NAT.

Jorropo commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,170 @@
+In the current version of DCUtR holepunching:
+
+- Performance is normal when a direct connection is established.
+- Performance is sub-optimal when someone connects to you, because all connections must go through a relay first, which adds an average delay of 5 seconds.
```
I'm not sure I understand this.

When you connect the connection signaling goes through the relay, thus have a ~5s latency for opening the connection.
Once it is open, you have a direct connection that have normal latency.

Jorropo commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,170 @@
+
+For a deeper dive into how holepunching works in Kubo, it's drawbacks, and more information on using DCUtR, see the [blog post](https://blog.ipfs.tech/2022-01-20-libp2p-hole-punching/).
+
+Because of these drawbacks, you may want to disable DCUtR holepunching and use another solution, like manual port forwarding. To disable DCUtR holepunching, set [`Swarm.EnableHolePunching`]( https://github.com/ipfs/kubo/blob/master/docs/config.md#swarmenableholepunching) to `false`:
```
Again, you don't need to disable it, it's trying to answer:
> It works (kinda), why would I want something better ?

Jorropo commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,170 @@
+   The default location for the config file is `~/.ipfs/config`. If you have set `$IPFS_PATH`, you can find your config file at `$IPFS_PATH/config`.
+   :::
+   
+1. Find the entry for `EnableHolePunching` in `Swarm` or create an entry for `EnableHolePunching` in `Swarm`, and set it to `false`:
```
See message above.

Jorropo reviewed (changes requested):
I belive that the last round, looks almost perfect :wink:

ElPaisano commented:
@Jorropo 

Regarding the DCUtR section issues: I rewrote it to (hopefully) clarify that
- it's enabled by default 
- it has some performance and reliability issues 
- setting up manual port forwarding will "disable" DCUtR and address the perf / reliability issues https://bafybeieha5qk2iwojsrpmtufuz7ke4efrnbba7lr4k7bnxjozab2v7uwja.on.fleek.co/how-to/nat-configuration/#use-dcutr-holepunching


Regarding your note that IPv6 does not always solve everything:
Updated IPv6 section with a note https://bafybeieha5qk2iwojsrpmtufuz7ke4efrnbba7lr4k7bnxjozab2v7uwja.on.fleek.co/how-to/nat-configuration/#enable-ipv6

Jorropo commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,132 @@
+
+As of Kubo v0.13, [DCUtR hole punching is enabled by default](https://github.com/ipfs/kubo/blob/master/docs/changelogs/v0.13.md#-relay-v2-client-with-auto-discovery-swarmrelayclient).
+
+DCUtR holepunching has various drawbacks and tradeoffs. Currently, the connection signaling goes through a relay, which can cause an average latency of 5 seconds when opening a connection. Once a direct connection is established, latency is generally normal. Additionally, DCUtR holepunching does not have a 100% success rate and can fail. For a deeper dive into how holepunching works in Kubo, it's ...
```
```suggestion
DCUtR holepunching has various drawbacks and tradeoffs. Currently, the connection signaling goes through a relay, which add an average latency of 5 seconds when opening a connection. Once a direct connection is established, latency is normal. Additionally, DCUtR holepunching does not have a 100% success rate and can fail. For a deeper dive into how holepunching works in Kubo, it's drawbacks, and more information on using DCUtR, see the [_Hole punching in libp2p - Overcoming Firewalls_ blog post](https://blog.ipfs.tech/2022-01-20-libp2p-hole-punching/). 
```

Jorropo reviewed (approved):
LGTM (except one slight wording about DCUtR sorry I'm being anoying)

TMoMoreau commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,132 @@
+
+### Enable manual port forwarding
+
+If your router does not support UPNP and/or IPv6, or you want better reliability and performance than what DCUtR provides, set up manual port forwarding. Complete the following steps to enable manual port forwarding:
```
Change "UPNP" to "UPnP", to be consistent with the rest of the doc.

ElPaisano replied:
Agreed

TMoMoreau commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,132 @@
 1. Reboot your IPFS node for the changes to take effect. Make sure to reboot the entire machine, not just the IPFS daemon.
 
-Assuming you're not trying to expose your daemon's API to the public internet, opening port `4001`/`tcp` should be sufficient.
+Now that you've opened a port from the internet to your internal Kubo node, update your Kubo configuration to set `Swarm.AppendAnnounce` as a list of addresses that other IPFS nodes will try to contact you at.
```
I think this paragraph could likely be removed completely. If I'm not mistaken, it is just an introduction to the next step, but the next paragraph already includes an introduction to the next step.

I think either remove it, or move it so that it follows the next header, and reword it to avoid repeating the same information.

ElPaisano replied:
Removing it make most sense to me

TMoMoreau commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,132 @@
+
+#### Update the Kubo Configuration
+
+In this step, you will update your Kubo configuration to set `Swarm.AppendAnnounce` as a list of addresses that other IPFS nodes will try to contact you at. This list lets other nodes on the network that the port forward you created in the previous step know that your node exists, and that these are the addresses at which you can be contacted at. To update your configuration:
```
Something went wonky with this paragraph.

"In this step, you will update your Kubo configuration to set `Swarm.AppendAnnounce` as a list of addresses that other IPFS nodes will try to contact you at. This list lets other nodes on the network (**know**) that the port forward you created in the previous step ~~know that your node~~ exists, and that these are the addresses at which you can be contacted at.

Something like that.

ElPaisano replied:
Your rewrite makes sense, I vote yes

TMoMoreau commented on docs/how-to/nat-configuration.md:
````diff
@@ -1,32 +1,132 @@
+    ],
+   ```
+
+Now that you've updated your Kubo configuration, restart Kubo so that your node is reachable, and check that the outside port is open.
````
I would suggest moving this line to follow the next header, as it is essentially an introduction to the next steps.

ElPaisano replied:
Agreed

TMoMoreau commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,132 @@
+#### Restart your Kubo node
+
+1. Reboot your IPFS node for the changes to take effect. Make sure to reboot the entire machine, not just the IPFS daemon. Once Kubo restarts, your node should be reachable. 
+1. Check that the outside port is open.
```
Consider adding an additional step to this, one that describes how the user would check that the outside port is open.

ElPaisano replied:
I agree. However, I don't think this should block adding the rest of this in. I'd vote push as is, and then we can update that step early next week

TMoMoreau reviewed (changes requested):
Requesting some changes to the style, formatting, etc.

Should be good to go after these are addressed.

ElPaisano commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,132 @@
 
-## IPv6
+:::callout
+While enabling IPv6 can fix many connection issues, enabling it may not address all of your connection issues. For example,  your router might support IPv6, but might still be running a firewall that blocks inbound connections.
```
Agreed

ElPaisano commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,132 @@
+# Configure NAT and Port Forwarding
 
-A common issue for new users running IPFS, especially on home networks, is that their IPFS node is _sitting_ behind a NAT (network address translation). It's common for people who run an IPFS node on their home network to have very long wait times or a high rate of request failure. This is because the rest of the nodes on the IPFS network have difficulty connecting to them through their NAT. The  ...
+A common issue for new users running IPFS Kubo, especially on home networks, is that their IPFS node is behind a [NAT (network address translation)](../concepts/glossary.md#nat) layer (such as a residential network router) experience very long wait times or high rate of request failure.
```
I like this rewrite, I vote yes

ElPaisano commented on docs/how-to/nat-configuration.md:
```diff
@@ -1,32 +1,132 @@
 ---
 
-# NAT configuration
+# Configure NAT and Port Forwarding
```
Agreed

ElPaisano commented:
> Requesting some changes to the style, formatting, etc.
> 
> Should be good to go after these are addressed.

@TMoMoreau nice review, I appreciate those catches 🔍 I agree with the majority of your changes. Take a look at individual comments.

TMoMoreau commented:
+1 for getting this in, and addressing the additional step next week.

TMoMoreau reviewed (approved):
Looks good, merge in when ready.
